extensions/lifecycle/lifecycleProducer/LifecycleProducer.js

```diff
@@ -0,0 +1,378 @@
+     * @param {String} accountCreds.accessKeyId - The account access key
+     * @param {String} accountCreds.secretAccessKey - The account secret key
+     * @return {AWS.S3} The S3 client instance to make requests with
+     */
+    _getS3Client(accountCreds) {
+        return new AWS.S3({
+            endpoint: this._s3Endpoint,
+            credentials: {
+                accessKeyId: accountCreds.accessKeyId,
+                secretAccessKey: accountCreds.secretAccessKey,
+            },
+            sslEnabled: this._transport === 'https',
+            s3ForcePathStyle: true,
+            signatureVersion: 'v4',
+            httpOptions: {
+                agent: new http.Agent({ keepAlive: true }),
+                timeout: 0,
+            },
+            maxRetries: 0,
+        });
+    }
+
+    /**
+     * Determine whether the given config should be processed.
+     * @param {Object} config - The bucket lifecycle configuration
+     * @return {Boolean} Whether the config should be processed
+     */
+    _shouldProcessConfig(config) {
+        if (config.Rules.length === 0) {
+            this._log.debug('bucket lifecycle config has no rules to process', {
+                config,
+            });
+            return false;
+        }
+        const { rules } = this._lcConfig;
+        // Check if backbeat config has a lifecycle rule enabled for processing.
+        const enabled = Object.keys(rules).some(rule => rules[rule].enabled);
+        if (!enabled) {
+            this.log.debug('no lifecycle rules enabled in backbeat config');
+        }
+        return enabled;
+    }
+
+    /**
+     * Process the given bucket entry, get the bucket's lifecycle configuration,
+     * and schedule a task with the lifecycle configuration rules, if
+     * applicable.
+     * @param {Object} entry - The kafka entry containing the information for
+     * performing a listing of the bucket's objects
+     * @param {Object} entry.value - The value of the entry object
+     * @param {String} entry.value.bucket - The bucket which has objects to
+     * perform lifecycle actions on
+     * @param {String} entry.value.owner - The canonical ID of the bucket owner
+     * @param {String} [entry.value.prefix] - An object prefix
+     * @param {String} [entry.value.keyMarker] - KeyMarker to list objects from
+     * @param {String} [entry.value.versionIdMarker] - VersionIdMarker to list
+     * versioned objects from
+     * @param {String} [entry.value.continuationToken] - ContinuationToken
+     * indicating that the list is being continued on this bucket
+     * @param {Function} cb - The callback to call
+     * @return {undefined}
+     */
+    _processBucketEntry(entry, cb) {
+        const { error, result } = safeJsonParse(entry.value);
+        if (error) {
+            this._log.error('could not parse bucket entry', { error });
+            return process.nextTick(() => cb(error));
+        }
+        const { bucket, owner } = result.value;
```

Comment:
I think the correct deconstruction is `const { bucket, owner } = result;`, because the kafka message object has no "value" field but "bucket" and "owner" are top-level fields of the message object.

Reply:
This exception occurred:
```
/home/scality/backbeat/extensions/lifecycle/lifecycleProducer/LifecycleProducer.js:177
        const { bucket, owner } = result.value;
                                        ^

TypeError: Cannot match against 'undefined' or 'null'.
    at LifecycleProducer._processBucketEntry (/home/scality/backbeat/extensions/lifecycle/lifecycleProducer/LifecycleProducer.js:177:41)
    at /home/scality/backbeat/node_modules/async/dist/async.js:4082:9
    at Object.process (/home/scality/backbeat/node_modules/async/dist/async.js:2330:17)
    at /home/scality/backbeat/node_modules/async/dist/async.js:2238:19
    at Immediate.<anonymous> (/home/scality/backbeat/node_modules/async/dist/async.js:119:16)
    at runCallback (timers.js:649:20)
    at tryOnImmediate (timers.js:622:5)
    at processImmediate [as _immediateCallback] (timers.js:594:5)
```

We should both fix the case, and also make the parsing of messages more resilient (i.e. do not crash but log an error and keep processing the next messages).

Reply:
I had seen this structure when publishing to a topic with a Kafka producer and needed to deconstruct this way. I will validate this one way or another as I'm adding tests.

Reply:
Okay, I will update accordingly.

Reply:
I think Bennett added a good try/catch for json parsing. For now, with the fix, I also check to make sure `bucket` and `owner` exist. If not, log error and return